_posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md

```diff
@@ -0,0 +1,75 @@
+---
+title: "Find definitions of Rails methods using source_location and bundle open"
+description: "Show me where the magic comes from."
+layout: article
+category: ruby
+image:
+  base: "2023/find-definitions-of-rails-methods-using-source_location"
+  alt: "Landsat maps in black and white"
+  credit: "USGS"
+  source: "https://unsplash.com/photos/white-and-black-abstract-painting-PuLsDCBbyBM"
+
+---  
+
+As a dynamic language it's important that Ruby provides many excellent debugging and introspection features.
+
+Finding the place a method or block of code is defined and reading its source code is essential for effective debugging and code comprehension. In Ruby, `source_location`, provides a powerful tool for retrieving the file and line number where a particular method or block is defined.
+
+
+## Explore Rails using…
+
+...the `#source_location` method.
+
+```ruby
+"Cars".method(:singularize).source_location
+#=> ["/.../activesupport-7.0.8/lib/active_support/core_ext/string/inflections.rb", 60]
+```
+
+Then open from the command line:
+
+```shell
+bundle open activesupport
+```
+
+Which opens the named gem in your editor of choice for you to explore.
+
+
+## Why?
+
+Use of `source_location` is invaluable when you're new to a application, re-exploring unfamiliar code or  trying to understand which gem is providing functionality you're using.
+
+Reading source code is a great way to learn. Reading battle-tested code like Rails, or other gems, even more so.
+
+Thanks to authors of the framework it even works for some of the meta-programmed "generated methods" in Rails. Methods on Active Record [associations that are generated](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/builder/association.rb#L103) using `class_eval` pass special syntax to enable the `source_location` lookup to still work (or you'd just see `(eval)` as the first result.
```

Comment:
Should the hyperlink include the 'class_eval' part?

Reply:
I think this paragraph is a little jumbled, especially towards the end?

Reply:
```suggestion
Thanks to authors of the framework `source_location` works for "magic" meta-programmed methods in Rails. Methods on Active Record [associations that are generated using `class_eval`](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/builder/association.rb#L103) pass special syntax to enable the lookup to still work. If it weren't for this when you called `source_location` you'd just see `(eval)` as the first result.
```